### PR TITLE
UPE Block Checkout

### LINKED ISSUE #2162: UPE working with the block checkout
So far all the effort for UPE has been using classic checkout. This issue is to get the UPE checkout flow working with the block checkout. Instructions for enabling the checkout blocks can be found [here](https://docs.woocommerce.com/document/using-the-new-block-based-checkout/).

### PR DESCRIPTION
**Pull request by bborman22**
Fixes #2162 

#### Changes proposed in this Pull Request

This PR adds the UPE to the block checkout experience. When UPE is enabled and the store is using the block checkout, UPE will now be loaded into the payment section. The block checkout makes use of the Stripe's React PaymentElement and so needed to update the `@stripe/react-stripe-js` package to the latest (1.4.1) this update also forced an update to the `@stripe/stripe-js` package to avoid any linting issues.

I kept the UPE block registration flow as independent as possible from the CC block registration flow in case there is ever a need to customize it without adding a lot of conditional checks. There is also the new UPE fields component to then use the PaymentElement. 

Some of the more notable details of the UPE fields component flow:
- The first was just that we need to create a payment intention before mounting the PaymentElement. I commented on the code here for any input on a better way to handle this just in case the approach I took isn't optimal. 
- Once mounted, there is an onChange event listener to check for the UPE form being completed.
- Once submitted, the onPaymentProcessing event will fire and check for the completed status and throw an error if not complete.
- In the classic checkout flow, we simulated a confirmPayment call to trigger the field errors. In this implementation I was unable to call the confirmPayment with an empty return_url as it caused it's own error. So the individual field errors will not be displayed unless the user has focused and blurred on them.
- Finally, the onCheckoutAfterProcessingWithSuccess is called after the server is updated. This part was tricky as we need to update the intent in case there was any price changes (different shipping method, coupon applied, etc). To update the intent required the order ID which was not available until this point. This leads to a slightly longer period between submitting the checkout and the final redirection from Stripe.

While working on this, we also got the feature from Stripe to pass in the billing details as needed and so that has been implemented here as part of the PR.

#### Testing instructions

- Start with a store with UPE enabled and Checkout blocks [enabled](https://docs.woocommerce.com/document/using-the-new-block-based-checkout/).
- Have 2 shipping methods, one flat rate (I used $10) and one free shipping.
- Add an item to the cart and select the flat rate shipping option from the cart.
- Confirm checkout loads properly and displays the UPE in the payments area.
- Try to submit without providing any payment details and confirm error is displayed.
- Provide only partial payment information, submit, and confirm error is displayed.
- Change the shipping method to the free shipping option, confirm the checkout total updates to reflect the change.
- Provide full payment information and submit
- Confirm that the checkout is successful 
- Confirm that the order was for the correct total (using free shipping)
- From the admin, view the transaction and confirm that customer details (name, email, address, etc) were correctly recorded for the transaction
- Run through the above steps using Euros and other payment methods (Giropay and Sofort)

Due to the update in Stripe JS packages, perform additional regression testing with: Classic CC checkout, Block CC checkout, Classic UPE checkout. Confirm all functionality appears unaffected. 

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: client/checkout/blocks/upe-fields.js
Comment (bborman22) [inline client/checkout/blocks/upe-fields.js:44]: Is there a better way to create the intention on component load or maybe even before that?

File: client/checkout/blocks/confirm-upe-payment.js
Comment (FangedParakeet) [inline client/checkout/blocks/confirm-upe-payment.js:22]:
```suggestion
	const name = `${ fields.billing_first_name } ${ fields.billing_last_name }`.trim();
```
Originally suggested by @frosso when I implemented this on the regular checkout, but I agree that this interpolation looks a little cleaner.

File: client/checkout/blocks/upe-fields.js
Comment (FangedParakeet) [reply client/checkout/blocks/upe-fields.js:44]: I'm not sure if there's a better React-y way to do this, but this seems to me like the appropriate place to keep this hook for this purpose.

File: client/checkout/blocks/upe-fields.js
Comment (FangedParakeet) [inline client/checkout/blocks/upe-fields.js:78]:
```suggestion
	// Once the server has completed payment processing, confirm the intent if necessary.
```
I feel bad even putting this as a suggestion, but I've seen it and now I can't stop seeing it, sorry.

File: client/checkout/blocks/upe-fields.js
Comment (FangedParakeet) [inline client/checkout/blocks/upe-fields.js:128]: Is there any reason why you only pass these specific fields, as the fields that the UPE expects we will collect ourselves? You are passing in the rest of the form fields [over here](https://github.com/Automattic/woocommerce-payments/blob/ff190b89037cac25b2ef488f743b9a08254c6fb1/client/checkout/blocks/confirm-upe-payment.js#L25-L42); for example, I think if you pass the `name` field as well, then Giropay will hide this field at checkout.

File: includes/class-wc-payments-upe-blocks-payment-method.php
Comment (FangedParakeet) [inline includes/class-wc-payments-upe-blocks-payment-method.php:52]: I guess I don't really mind the presence of this class for now, but could you just have inherited from the `WC_Payments_Blocks_Payment_Method` class, since it appears this is the only part that differs. `UPE_Payment_Gateway::GATEWAY_ID` and `WC_Payment_Gateway_WCPay::GATEWAY_ID` both return the same id string, so that seems a little redundant inside the `initialize()` function. But please correct me if I'm being short-sighted and missing anything here!

File: client/checkout/blocks/upe.js
Comment (FangedParakeet) [inline client/checkout/blocks/upe.js:43]: Is there no way to pull these fields from the Gateway class instead of repeating them here? While I am criticising this, ironically this may actually work out in our favour, if we wanted to keep this title dynamic dependent on enabled payment methods, because I think we should be able to pull enabled payment methods using `getConfig( 'paymentMethodsConfig' )` at this point. Let me know if you prophetically had this in your mind over here!

**Review by FangedParakeet — CHANGES_REQUESTED** [review]
I tested this blocks checkout pretty thoroughly and the previous blocks and UPE checkout a little more lightly. I've left a few suggestions for little code things here and there in other comments.

I did notice one rather large thing that was not working for me, but perhaps there is a reason for it beyond my cognition. I noticed that whenever I was checking out with this Blocks UPE checkout, I could not see any of my saved payment methods anywhere at the checkout. I couldn't see the checkbox to save a new payment method either, though both of these items appeared on the classic Blocks checkout. Consequently, I was able to complete a checkout for a subscription product, but I'm not sure if it correctly saved my payment method, since I could not find the new payment method I used anywhere in my saved payment methods on my account.

Also, on the subscription side, I was unable to complete a checkout for a subscription product with a free-trial. The Blocks checkout threw an error that wasn't handled within the Promise and lit up my console crimson instead. The error itself  ("Error: This value must be greater than or equal to 1.") suggests that the checkout is creating a payment intent with an amount of 0, instead of a setup intent; I have a sneaking suspicion that if you merge the latest changes committed into develop, some of this problem will resolve itself, but I think you will still have to add a little extra logic to `confirmSetup` instead of `confirmPayment`.

There are quite a few cases to cover here, so let me know if you planned on handling saved payments/subscriptions in a separate issue or if you'd prefer to get it all done at once. I didn't have any issues with the regular checkout for simple products!

File: client/checkout/blocks/upe-fields.js
Comment (bborman22) [reply client/checkout/blocks/upe-fields.js]: Nope, just hadn't gone back to it after the initial set up. These should all be removed from the fields now with the latest push.

File: includes/class-wc-payments-upe-blocks-payment-method.php
Comment (bborman22) [reply includes/class-wc-payments-upe-blocks-payment-method.php:32]: Good call, I was really torn on this one, so with your nudge, I went the route of inheritance now with the latest push.

File: client/checkout/blocks/upe.js
Comment (bborman22) [reply client/checkout/blocks/upe.js:43]: So there was a little bit of foresight going in to this decision as you mentioned since I knew we wanted to alter these titles anyway. I also was just using the existing CC implementation as a starting point and went with their approach. With that in mind do you think it would be best to keep this as is and address the dynamic title in #2433?

Comment (FangedParakeet) [reply client/checkout/blocks/upe.js:43]: I think let's leave it like this; I was just thinking out loud with my comment and as I did so, eventually came around to this same reasoning. I think #2433 will probably end up overwriting this: it'll still probably be nice to pull the optional values here from the gateway config, but that can be taken care of properly in the separate issue.

**Comment by bborman22** [comment]
I have the save payment method flow in a good spot locally, but having #2499 merged first would allow me to avoid pushing changes with a possible failed scenario. I'm going to hold off on this PR until that is merged and then will be able to close this one out with the full save payment methods functionality.

**Comment by bborman22** [comment]
I have added in the saved payment methods and free trial subscription features now with some details below.

For the saved payment methods, the process for displaying and allowing them to be selected was straightforward. The ability to save them was also pretty simple just using the `shouldSavePayment` property [passed in to the element](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/docs/extensibility/payment-method-integration.md#props-fed-to-payment-method-nodes). The only tricky part was disabling the ability to save a payment method that doesn't allow it. To address this I added an error message when a customer attempts to save a payment method that can't be saved. I will create an enhancement issue for this as it's not the greatest solution, but works within our current constraints.

For the free trial subscriptions, it was just a matter of checking the `payment_needed` value in the checkout response. I kept all the logic within the `confirmUPEPayment` function since it is all the same up to the confirm call which I wrapped in a conditional. There is one known issue in that when purchasing a subscription the save payment method checkbox is not automatically checked for the customer. This matches the current CC gateway implementation and may just be a limitation of the checkout block functionality at this time. This feature would play in with the same enhancement issue discussed above (essentially the ability to modify the `showSaveOption` after loading).

All other features of the UPE block checkout should now be implemented and functioning as expected!

File: client/checkout/blocks/upe-fields.js
Comment (FangedParakeet) [inline client/checkout/blocks/upe-fields.js:56]: There's an uncaught exception here when `api.createIntent` throws an error. This error just gets dropped in the console and there's nothing displayed in the UI, which is a little confusing (this happened to me, because I had the wrong currency selected--which I know is being fixed in another issue). Maybe do you fancy wrapping this in a `try/catch` or something and returning the error to the UI?

**Review by FangedParakeet — COMMENTED** [review]
@bborman22, nice one on this! I ran through all the regular flows and everything appeared to be working for the most part as expected. 

I think we can leave saved payments the way they are now, simply showing that error message when the payment method doesn't support being saved; I presume there was something painfully fiddly about reaching that element, so fine leaving that as a future enhancement.

What's the score on subscriptions? I was able to checkout with a regular subscription and free-trial subscription product in a way that appeared successful, though I'm not sure whether my payment method was correctly saved. Anyways, it appears to me as if this is working or at least it's working as well as it would for the regular WCPay Blocks checkout, so I'd give this the green light. 

There are two tiny issues I noticed with uncaught exceptions while creating and updating payment intents--can you just have a quick look at those and then I'll gladly approve. Cheers, mate!

File: client/checkout/blocks/upe-fields.js
Comment (bborman22) [reply client/checkout/blocks/upe-fields.js]: Great catch, added this in the latest updates.

File: client/checkout/blocks/upe-fields.js
Comment (bborman22) [reply client/checkout/blocks/upe-fields.js]: I've got a try/catch on this now. However, displaying the error message proved a little more difficult than I expected as it seems you can only "bubble up" errors during the defined events, which would require the customer to try to submit the form first. I found a workaround that isn't super pretty, but at least gets the message across clearly.

**Review by FangedParakeet — APPROVED** [review]
Thanks for the changes here: I think the way you handled the exceptions via React is pretty acceptable and failing the intent calls displayed the expected error for me. Everything else is still working as expected, as far as I could tell.

This code L. G. T. M. and with great pride I give you the green light.

**Pull request merged**
